### DISCUSSION TITLE
Custom OAuth providers

### DISCUSSION THREAD

**Discussion by LoriKarikari** [supabase/supabase#417] [Questions]
Hi,

Is it possible to add custom OAuth providers if you want to use one that isn't built-in yet?

**Comment by kiwicopple** [comment]
Hey @LoriKarikari  - do you mean another provider like "Twitter", "Facebook" etc? We're working on these now. Are there any providers in particular that you're looking for?

**Reply by valerius21** [reply]
Heya @kiwicopple, I am looking for something similar like @LoriKarikari. I want to intergrade the discord oath provider for example. Is there an abstract class from which you could build your own provider Handler? (Or something similar?) Or are we limited to use the given Providers, like Github, Google, etc.?

**Reply by LoriKarikari** [reply]
I currently don't need a specific provider, I was just checking out Supabase and was wondering if we were limited to the built-in providers 😄.  As @valerius21 said, it would be nice to be able to create your own handler for whatever OAuth provider you want to use.

**Reply by kiwicopple** [reply]
You can always use your own Auth provider - ultimately Supabase is just a Postgres database, so you can implement Auth however you would usually implement it (eg, in your middleware). Our auth solution just makes it easier so that you don't have to do it yourself.

If you wanted to still use Row Level Security, then you would need to add the user's JWT to the client library:

```js
const headers = { Authorization: 'Bearer USER_JWT' }

const supabase = createClient( "https://REF.supabase.co",  "KEY", { headers });
```

Hopefully I understood your questions correctly - let me know if I misunderstood!

**Reply by valerius21** [reply]
Hey @kiwicopple, thank you dearly for the fast reply! I am going to try it out in the next days!

**Reply by MarvinMiles** [reply]
Would be nice to have Discord provider built-in as well 🙃

**Reply by kiwicopple** [reply]
Feel free to request over at `supabase/gotrue` repo :)

**Reply by eifo** [reply]
Trying to add authorization header for use with next api routes for RLS still not getting it right what am I missing?

const access_token = JSON.parse(
  typeof window !== 'undefined' ? localStorage.getItem("supabase.auth.token") : null
)?.currentSession?.access_token

const headers = {Authorization: `Bearer ${access_token}`}
const supabase = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL,process.env.NEXT_PUBLIC_SUPABASE_KEY,headers);

**Reply by kiwicopple** [reply]
We have an open issue for this one @eifo : https://github.com/supabase/gotrue-js/issues/34#issuecomment-762038518

**Reply by fangmarks** [reply]
> Are there any providers in particular that you're looking for?

Personally, I'd love to see support for Discord and Twitter

**Reply by nascode** [reply]
Is there any complete tutorial or docs on how to implement custom provider?

In my hobby project, I would like to try using Discord login, then check if user has joined certain Discord server.

**Reply by warflash** [reply]
Hey,
wondering if there´s any update on this topic or if someone got a guide/docs on how to properly implement 3rd party auth of discord/twitter? Would be much appreciated as docs on custom use of supabase auth still seem very hard to come by.

**Reply by rkhatkhede** [reply]
Most of my clients have their own oauth solutions. There are so many ways of implementing the same thing but some handlers to implement oauth2 the right way which is also integrated with supabase environment like UI and User management will be super cool. maybe something like https://auth0.com/docs/connections/social/oauth2

**Reply by JanRuettinger** [reply]
Is there a tutorial or a guide on how to implement your own oauth provider with supabase?

**Reply by baptisteArno** [reply]
I also need this... Did you find something guys?

**Reply by baptisteArno** [reply]
As soon as I get the access token and refresh token from the provider, how can I create a user? @kiwicopple
(I'll post a complete guide on how to authenticate with a custom provider as soon as I get this right 👍)

```ts
const userInfo = await spotifyApi.getMe();
const user = {
    aud: "authenticated",
    exp: 1615824388,
    sub: userInfo.body.id,
    email: userInfo.body.email,
    app_metadata: {
      provider: "spotify",
      access_token: data.body["access_token"],
      refresh_token: data.body["refresh_token"],
    },
    user_metadata: {},
    role: "authenticated",
};

// INSERT IN auth.users
```

**Reply by liaujianjie** [reply]
@baptisteArno Seems like Spotify OAuth might be landing soon: https://github.com/supabase/gotrue/pull/245

**Reply by marc-on-github** [reply]
Have you found a solution?

**Reply by gdaybrice** [reply]
@kiwicopple is there any workaround you might recommend for now to byo an oAuth provider. Any API in GoTrue we can use or would it work to handle the custom provider callbacks with edge functions and write straight to the auth table?

**Reply by xiakunhou** [reply]
@baptisteArno Have you found a solution to create user for customized auth provider?

**Comment by SealOfTime** [comment]
I am curious about the same question. 

There are dozens of different oAuth providers, that may be local to certain countries or enterprise spheres: Sber, VKontakte (aka Facebook-competitor) and Yandex (aka Google-competitor) in Russia to begin with.  

I can't see even the immensely active community of Supabase to implement all of them into the default distributipn. So the question arises, is there a possibility we would be able to define our own providers in a friendly UI, or at least as a some sort of a config? 
Moreover, there are different authentication protocols specific to certain applications like LDAP for internal corporative systems, that my be desired by potential customers.

Would appreciate, if you would consider either giving the administrator ability to add custom oAuth providers or at least adding a chapter to the documentation about implementing custom authentication service and integrating that with Supabase's authorization. 

Looking forward to your response!

**Reply by gracew** [reply]
Information on how to use custom auth AND RLS is available in various Supabase docs but it took me a few hours last night to pull it together. So I wrote up [this guide](https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d), hope it helps some of you!

TL;DR:
1. Obtain the Supabase JWT secret
1. When the user authenticates, set a cookie containing a signed JWT from the server
1. Call setAuth() from the client to make requests as the user
1. Use the JWT data in either your RLS policy or in column default values

**Reply by SealOfTime** [reply]
Thanks a lot! That's definitely one way to do it!

**Reply by humphd** [reply]
@gracew just wanted to say 'thank-you' for this post, it was *exactly* what I needed.  We're using it to implement our custom auth.  You're a hero!

**Reply by Aldo111** [reply]
Thanks for sharing the guide, Grace! I was curious what might be the appropriate way to insert that data into `auth.users` (or is it not a good idea to do that with custom auth providers/jwts). I ultimately want to make sure all my registered users are in one place in the auth side of the database.

**Reply by OliverRhyme** [reply]
How can we handle token refresh with this?

**Reply by olee** [reply]
Same question: How to generate the correct refresh tokens this way?

**Reply by xiakunhou** [reply]
@gracew So in your solution, the auth tables would contain no user info(for example, the user_id would not be genreated), while only depends on the customized jwt playload to do RLS in posrgresql?

**Reply by Revadike** [reply]
> Information on how to use custom auth AND RLS is available in various Supabase docs but it took me a few hours last night to pull it together. So I wrote up [this guide](https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d), hope it helps some of you!
> 
> TL;DR:
> 
>     1. Obtain the Supabase JWT secret
> 
>     2. When the user authenticates, set a cookie containing a signed JWT from the server
> 
>     3. Call setAuth() from the client to make requests as the user
> 
>     4. Use the JWT data in either your RLS policy or in column default values

Would you mind updating your guide, now that `setAuth` is deprecated?
https://supabase.com/docs/reference/javascript/release-notes#deprecated-setauth

**Comment by halfong** [comment]
Consider make auth support for **Wechat**.

**Reply by kchojhu** [reply]
@gracew You are awesome! Looks like that is a great way to integrate with any authentication layer.  I hope supabase would have official support for custom auth (maybe that can be done with functions)

**Reply by wwz223** [reply]
have you found a solution

**Comment by magick93** [comment]
Please consider Keycloak!

**Reply by fangmarks** [reply]
Pretty sure Keycloak got support in the latest Launchweek!

**Reply by fangmarks** [reply]
Yup

https://supabase.com/docs/guides/auth/auth-keycloak

**Comment by modyabhi** [comment]
Implement openID 2.0 for steam web authentication

**Reply by Revadike** [reply]
[I attempted a PR for Steam provider support](https://github.com/supabase/auth/pull/1919)
However, I think the main problem is Supabase's reliability on user emails, so I am not sure if this will (fully) work.

**Comment by blanklob** [comment]
Can you please add Shopify ?

**Comment by kris** [comment]
Instagram, TikTok, Pinterest support? I'd love if we could get some sort of documentation on how to add custom providers.

**Reply by sambowenhughes** [reply]
+1 for Tiktok

**Reply by n0rmanc** [reply]
+1 for Tiktok

**Reply by augusteo** [reply]
+1 for Tiktok

**Reply by bhkangw** [reply]
Anotherrrrr +1 for Tiktok pls!

**Reply by pateljoel** [reply]
+1 for these OAuth login options, thanks @kris for mentioning this.

@kiwicopple these networks *combined* have a total **4.282 BILLION** monthly active users, this would be massive for supabase to support these OAuth login options and will increase adoption of builders using supabase and more users using the apps on your platform.

Please consider TikTok, Instagram and Pinterest login.

**Reply by jonatandorozco** [reply]
+1 for tiktok

**Comment by Jonevs** [comment]
Atlassian, please. Needed it for my Jira Integration App

**Comment by imjamesb** [comment]
Please add Vipps authentication. https://vipps.no

This is a Norwegian payment service.

**Comment by taujin** [comment]
Looking to integrate LINE oauth. Used pretty much all of Taiwan/Japan. Will there be a way to add OAUTH2 providers not on the list?

**Reply by Antiokh** [reply]
I'll upvote. So few plugins supporting this very popular Asian network.

**Reply by snak2e-tech** [reply]
Agreed. LINE is also used in a lot of Southeast Asia too. If you counted by the numbers... it probably is just as popular as telegram. Probably more so.

**Comment by jsmith212** [comment]
Interested in using Authentik for a self-hosted solution. Not sure if this is possible yet

**Reply by japarmedi** [reply]
I think SAML 2.0 integration is possible with Authentik and Supabase

**Comment by bigbizze** [comment]
would love an integration with quick books online https://developer.intuit.com/app/developer/qbo/docs/develop/authentication-and-authorization/oauth-2.0

**Comment by hxtnv** [comment]
Vk and Steam support would be awesome

**Reply by adambuchweitz** [reply]
+1 for Steam

**Reply by Hydra566** [reply]
VK and Yandex

**Reply by Revadike** [reply]
[I attempted a PR for Steam provider support](https://github.com/supabase/auth/pull/1919)
However, I think the main problem is Supabase's reliability on user emails, so I am not sure if this will (fully) work.

**Comment by Herklos** [comment]
Could you please add Binance? https://developers.binance.com/docs/login/introduction

**Comment by sandervspl** [comment]
I would love to see a provider for Battle.net!

https://develop.battle.net/documentation/guides/using-oauth

**Comment by libratiger** [comment]
looks for this too

**Comment by rohanliston** [comment]
It would be great for a generic OAuth2/OIDC adapter to be prioritised, as it would provide support for countless authentication providers rather than having to wait for specific providers to be added one-by-one. Is this on the roadmap?

**Reply by j-o-sh** [reply]
Even further: This would enable us to use custom _pseudo_ auth providers in development on localhost. Switching users in a multi-user app, while developing on localhost is currently a real pain and could be so much more elegant if we could just implement our own custom pseudo auth provider.

something like:
`config.toml`
```toml
[auth.external.custom.foobar]
enabled = true
client_id = "123"
secret = "donttellanyone"
redirect_uri = "http://localhost:54321/auth/v1/callback"
url = "http://localhost:9999"
```

**Reply by dtinth** [reply]
I found a hacky solution: Use Keycloak as authentication provider (which is based on OAuth 2.0) and create a custom API that imitates Keycloak’s API paths and behavior (there are only 3 endpoints to implement). More details in this comment: https://github.com/orgs/supabase/discussions/6547#discussioncomment-9049664

**Comment by rickylabs** [comment]
Any news on this ? Much anticipated feature for our company !

**Comment by starantini** [comment]
Hey guys, 
   anyone know if when calling this sign in with Id method, are the providers the same as the oAuth provider method. I've tried calling this method with provider set to slack. This method doesn't seem to recognize slack. 

 `const { data, error } = await supabase.auth.signInWithIdToken({
  provider: 'google',
  token: 'your-id-token'
})`

Error in browser: AuthApiError: Custom OIDC provider "" not allowed

I've confirmed that my id token is correct using jwt.io, is this a case that the list of providers for this method is not the same as the oAuth providers? Any work arounds? I can't use setSession method because I'm not able to get a refresh token and access token, so this was my work around to that. 

Appreciate any ideas here as I've been grinding away at this for days trying to include as many providers as possible.

**Reply by StolteX** [reply]
I have the same error message with github...

**Comment by WanChengCheng** [comment]
Any updates on this?

**Comment by hxtnv** [comment]
To all those still coming here, I wouldn't hope for much - I emailed support back in march of this year and they said it is "unlikely" that more individual providers will be added. Here's the full response:
![image](https://github.com/supabase/supabase/assets/39811538/a516cfdb-117d-42c0-8f69-ba47a77995a7)

**Comment by all-mute** [comment]
+1 for Yandex and VK

**Comment by markadrake** [comment]
One of the things that attracted us to Supabase was the promise to make authentication easy(er). While we would still benefit today by moving to Supabase, we cannot because Supabase does not include our "social providers." This includes Xbox, Playstation, and Nintendo.

I kindly request new documentation to cover "bringing your own" authentication provider.

Thanks for your hard work and excellent product,

**Comment by lewinskimaciej** [comment]
I found this thread and also wanted to chime in, that we assumed this is possible with Supabase, only to find out it's not. We have our partners with their own OAuth2 identity solutions, branded and customized so we cannot bring their users into our platform easily because we can't integrate with them. This seems like an incredibly basic and required feature to have at this point.

**Reply by dtinth** [reply]
Possible workaround for now, until Supabase implements support for generic OAuth2/OIDC provider: https://github.com/orgs/supabase/discussions/6547#discussioncomment-9049664

**Reply by lewinskimaciej** [reply]
Hey, yeah I saw this and I like the idea. The only issue is that you can have only one provider of each type right? This limits us on partners we can work with because we'll only be able to create one keycloak provider.

**Reply by dtinth** [reply]
In this case maybe you can install Keycloak and use it as an [identity broker](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker) instead of Supabase.

**Reply by lewinskimaciej** [reply]
Yeah I'm considering this, we're actually just keen on dropping Supabase entirely in favor of something like Ory that doesn't make that an issue, as our use case might require us to add multiple custom oauth providers in the future to our login solution.

**Comment by gdaybrice** [comment]
Looks like they will start working on this in September, so probably a good
6 months away. I’ve been looking at alternatives and Auth0 also supports
custom providers but would rather keep everything in supabase.

**Reply by RobbyUitbeijerse** [reply]
Hey @blechatellier, any chance you could link the information regarding starting this feature in September? We're contemplating on moving away from Supabase, but if the waiting time is ~5-6 months, we might just not do that - depending on what's in store.

@J0 - would love to hear your input - is this in the works? And will we be able to eg register different providers with the same provider type? the solution is currently quite limited, but Supabase offers us what we want in terms of experience - so we would rather stay :)

**Reply by gdaybrice** [reply]
@RobbyUitbeijerse second hand information from a convo I had on  Twitter https://x.com/stojaaan/status/1795421822114005163?s=46&t=zIzrZmJzfRoF4kjsueapFA

**Reply by RobbyUitbeijerse** [reply]
Thanks @blechatellier! Let's see...

**Reply by RobbyUitbeijerse** [reply]
@hf Hello there! Is this still planned in to be started in Sept?

**Reply by dabbijak** [reply]
@hf  Any update on this?

**Comment by japarmedi** [comment]
I think you guys need to build your own Identification Provider and integrate it with Supabase through SAML 2.0 integration

**Comment by C5G6M7** [comment]
Hi all,

If supabase wanted to add this feature as simple solution would be to add a migration script that adds a "Provider" table to the auth schema in the database, then update the auth/internal/api/external.go function which currently looks like this:

`
func (a *API) Provider(ctx context.Context, name string, scopes string) (provider.Provider, error) {
	config := a.config
	name = strings.ToLower(name)

	switch name {
	case "apple":
		return provider.NewAppleProvider(ctx, config.External.Apple)
	case "azure":
		return provider.NewAzureProvider(config.External.Azure, scopes)
	case "bitbucket":
		return provider.NewBitbucketProvider(config.External.Bitbucket)
	case "discord":
		return provider.NewDiscordProvider(config.External.Discord, scopes)
	case "facebook":
		return provider.NewFacebookProvider(config.External.Facebook, scopes)
	case "figma":
		return provider.NewFigmaProvider(config.External.Figma, scopes)
	case "fly":
		return provider.NewFlyProvider(config.External.Fly, scopes)
	case "github":
		return provider.NewGithubProvider(config.External.Github, scopes)
	case "gitlab":
		return provider.NewGitlabProvider(config.External.Gitlab, scopes)
	case "google":
		return provider.NewGoogleProvider(ctx, config.External.Google, scopes)
	case "kakao":
		return provider.NewKakaoProvider(config.External.Kakao, scopes)
	case "keycloak":
		return provider.NewKeycloakProvider(config.External.Keycloak, scopes)
	case "linkedin":
		return provider.NewLinkedinProvider(config.External.Linkedin, scopes)
	case "linkedin_oidc":
		return provider.NewLinkedinOIDCProvider(config.External.LinkedinOIDC, scopes)
	case "notion":
		return provider.NewNotionProvider(config.External.Notion)
	case "spotify":
		return provider.NewSpotifyProvider(config.External.Spotify, scopes)
	case "slack":
		return provider.NewSlackProvider(config.External.Slack, scopes)
	case "slack_oidc":
		return provider.NewSlackOIDCProvider(config.External.SlackOIDC, scopes)
	case "twitch":
		return provider.NewTwitchProvider(config.External.Twitch, scopes)
	case "twitter":
		return provider.NewTwitterProvider(config.External.Twitter, scopes)
	case "workos":
		return provider.NewWorkOSProvider(config.External.WorkOS)
	case "zoom":
		return provider.NewZoomProvider(config.External.Zoom)
	default:
		return nil, fmt.Errorf("Provider %s could not be found", name)
	}
}`

and change the default case at the end to call a method that checks for the existence of a custom provider in the database and return that as a new provider object if it exists.

**Comment by nathangroblewski** [comment]
Unfortunately, this might force me to move off Supabase, which I truly don't want to do. Is there any clarity at all that can be provided? Is it still on the RoadMap?

**Comment by daveycodez** [comment]
Lol hello?

**Comment by lema-founders** [comment]
Does anyone know how to implement IG login using custom OAuth? Very confused on what should the callback URLs on Supabase and / IG should be

**Comment by jesKo-dev** [comment]
Shopify integration would be nice. Connecting stores information and data to Supabase in order to use it for further projects.

**Comment by antonlvovych** [comment]
Any updates on this? It’s unbelievable that we still don’t have such a basic feature…

**Reply by miketaheny** [reply]
Completely agree.  Decided I have to move away from Supabase.  Assumed this was going to be an option, spent a ton of time investing in the platform and now poof.  wasted

**Reply by antonlvovych** [reply]
Same. Switched to auth.js because of this

**Reply by k-nearest-neighbor** [reply]
@antonlvovych Hey was it easy to switch? 

I see there is https://authjs.dev/getting-started/adapters/supabase?framework=express

Do you use row level security?

**Reply by Antiokh** [reply]
Also interested

**Comment by asimaranov** [comment]
Same, extremely need VK and Yandex authorizations that are presented in better-auth

**Reply by Antiokh** [reply]
And also Gosuslugi? =)

**Reply by georgiev-anton** [reply]
and MAX messanger ? 😂

**Comment by lordlothar99** [comment]
What about Firebase, now that it's possible to use Firebase as a 3rd party auth provider ?
https://supabase.com/docs/guides/auth/third-party/firebase-auth

**Comment by Alexandre-Chapelle** [comment]
I appreciate the incredible work everyone has been doing on this project it’s a ver solid and easy to use solution to a set of very complex problems, and it’s been a joy to use (mostly).
However, this feature is critically important to me and, I'm sure, to many other teams and developers using supabase. I do understand that this feature is not top priority, but it is very important for us. If this feature is something the team is not planning to work on in the following months, unfortunately we'll have to switch.

Thanks again for all the hard work, and I'd love to hear any updates or timelines regarding this feature.

**Reply by bdjafer** [reply]
same for us!

**Reply by coopbri** [reply]
Hey! What are you planning on switching to? Custom OIDC is a dealbreaker for us too

**Reply by Alexandre-Chapelle** [reply]
> Hey! What are you planning on switching to? Custom OIDC is a dealbreaker for us too

Hey! For us, keycloak has always been on our radar, it was actually one of our first considerations. However, its operational complexity (setup, scaling, maintenance) didn't align with our needs when we initially adopted supabase. Now, though, the lack of custom OIDC support forces us to reevaluate.

Switching to another baas (firebase, etc.) felt counterproductive, we'd just end up in the same situation in a few month, dependent on someone else's roadmap for critical features. While self-hosting supabase could work technically (we've looked into modifying GoTrue for extra providers), it defeats the purpose of using a managed service in the first place. If we're going to invest in infrastructure overhead, we'd rather pick a battle-tested, extensible solution like keycloak, even if it means sacrificing some of supabase's cool features (storage, functions, etc.).

Just to be clear: supabase has been doing very good for our auth workflows, and this isn't a reflection on the project or the team (y'all are doing incredible work). But since we're only using supabase for auth/oauth, the cost-benefit calculus shifts. keycloak, despite its steeper learning curve and a few other drawbacks, feels like the best option right now.

Would still love any updates on oidc timelines, though, if priorities shift, we'd absolutely revisit supabase.

**Comment by ralienpp** [comment]
Would it be possible to use Authentik, while configuring Supabase to use Keycloak? The idea is to make Authentik pretend that it is Keycloak; given that they are supposed to be using a standardized protocol - shouldn't it be irrelevant what server one runs, as long as it implements the protocol correctly?

**Comment by sdelanos** [comment]
Any updates about adding custom auth providers ?

**Comment by Antony-Jia** [comment]
I think https://github.com/trailbaseio/trailbase this database is better then supabase on cutom OAuth
